src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,321 @@
+        {
+            if (SymbolIsStaticAndAbstract(propertyOrMethodSymbol, requiresPreviewFeaturesSymbols))
+            {
+                // TODO: This needs to be tagged only on interfaces. A static abstract method on an abstract class is illegal and we shouldn't tag such cases.
+                if (runtimeFeatureType != null && HasPreviewAttribute(runtimeFeatureType, requiresPreviewFeaturesSymbols))
+                {
+                    requiresPreviewFeaturesSymbolsToUsageType.GetOrAdd(propertyOrMethodSymbol, PreviewFeatureUsageType.StaticAbstract);
+                    return true;
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsImplementationOfAnyImplicitInterfaceMember(out ISymbol baseInterfaceMember))
+            {
+                if (HasPreviewAttribute(baseInterfaceMember, requiresPreviewFeaturesSymbols) || ContainingTypeIsPreview(baseInterfaceMember, requiresPreviewFeaturesSymbols))
+                {
+                    return true;
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsOverride)
+            {
+                ISymbol? overridden = propertyOrMethodSymbol.GetOverriddenMember();
+                if (overridden != null && HasPreviewAttribute(overridden, requiresPreviewFeaturesSymbols))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            ISymbol symbol = context.Symbol;
+            if (symbol == null)
+            {
+                return;
+            }
+
+            // Don't report diagnostics on types that are marked as Preview
+            if (HasPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return;
+            }
+
+            bool reportDiagnostic = false;
+
+            // Look for attributes on base classes that this type may extend
+            if (symbol is ITypeSymbol typeSymbol)
+            {
+                reportDiagnostic = TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType);
+            }
+
+            if (!reportDiagnostic)
+            {
+                if (symbol is IMethodSymbol || symbol is IPropertySymbol)
+                {
+                    reportDiagnostic = PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType);
+                }
+            }
```

Comment:
This would be more readable as a switch expression:

```c#
var reportDiagnostic = symbol switch
{
   ITypeSymbol typeSymbol => TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType);
   IPropertySymbol _ or IMethodSymbol _ => PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType);
   _ => false;
}